File: 경주로건설.java
package hello.login;

import java.util.*;

public class Main {

    public static int n;

    public static int[] board = new int[15];
    public static int[][] map = new int[26][26];
    public static boolean[][] visited = new boolean[26][26];

    public static int[] dx = {1, -1, 0, 0};
    public static int[] dy = {0, 0, 1, -1};
    public static int limit;

    public static boolean is_prime(int n)
    {
        if(n<=1)
            return false;
        for(int i=2;i<=Math.sqrt(n);i++)
        {
            if(n%i==0)
                return false;
        }
        return true;
    }

   

    public static int solution(int[][] board) {
        int answer = 0;

        int len = board.length;
        for(int i=0;i<len;i++)
        {
            for(int j=0;j<len;j++)
            {
                map[i][j] = board[i][j];
            }
        }

        Queue<Path> q = new LinkedList<>();
        q.add(new Path(0, 0, 0)); //0,0 시작
        while(!q.isEmpty())
        {
            Path it = q.poll();
            int x = it.getX();
            int y = it.getY();
            int cost = it.getCost();
            for(int i=0;i<4;i++)
            {
                int nx = x+dx[i];
                int ny = y+dy[i];
                if(nx>=0 && nx<limit && ny>=0 && ny<limit)
                {
                    if(!visited[nx][ny])
                    {
                        visited[nx][ny] = true;
                        //코너?
                        
                        //직선도로
                    }
                }
            }
        }



        return answer;
    }

    static class Path
    {
        private int x;
        private int y;
        private int cost;

        public Path(int x, int y, int cost) {
            this.x = x;
            this.y = y;
            this.cost = cost;
        }

        public int getX() {
            return x;
        }

        public void setX(int x) {
            this.x = x;
        }

        public int getY() {
            return y;
        }

        public void setY(int y) {
            this.y = y;
        }

        public int getCost() {
            return cost;
        }

        public void setCost(int cost) {
            this.cost = cost;
        }
    }





    public static void main(String[] args) {

    }

}

File: 경주로건설.cpp
#include <string>
#include <vector>
#include <queue>
#include <tuple>
#include <iostream>


using namespace std;
bool visited[26][26][4];
vector<vector<int>> maps;
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};
int len;
int minCost = 999999;

void bfs(int x, int y, int dir, int cost)
{
    queue<tuple<int, int, int, int>> q;
    q.push({x, y, dir, cost});
    visited[x][y][0] = visited[x][y][1] = visited[x][y][2] = visited[x][y][3] = true;
    while(!q.empty())
    {
        auto it = q.front();
        q.pop();
        int px = get<0>(it);
        int py = get<1>(it);
        int pDir = get<2>(it);
        int pCost = get<3>(it);
        
        if(px==len-1 && py==len-1)
        {
            minCost = min(minCost, pCost);
        }
        
        for(int i=0;i<4;i++)
        {
            int nx = px+dx[i];
            int ny = py+dy[i];
            int nDir = i;
            int nCost = pCost;
            
            if(nx<0 || nx>=len || ny<0 || ny>=len || maps[nx][ny]==1)
                continue;
            
            if(pDir==-1)
                nCost+=100;
            else if(pDir==nDir)
                nCost+=100;
            else
                nCost+=600;
            
            if(!visited[nx][ny][nDir] || maps[nx][ny]>=nCost)
            {
                maps[nx][ny] = nCost;
                visited[nx][ny][nDir] = true;
                q.push({nx, ny, nDir, nCost});
            }
        }
    }
}


int solution(vector<vector<int>> board) {
    int answer = 0;
    maps = board;
    len = board.size();
    bfs(0, 0, -1, 0);
    answer = minCost;
    
    return answer;
}
